Type mysql dialect

Polandia94 opened the pull request:
Tipyng of mysql dialect.

possible breaking changes:

- on `visit_sequence`, `visit_true` and `visit_false` I changed the param names to be the same tht in the super class.
- Removed  `tostring` on `array`. `tostring` is deprecated since Python 3.2 and removed in Python3.9 so i guess we can remove that option altogheter.
- create_connect_args on aiomysql, added an unused param _translate_args, only to have same signature as super class.
-  delete _check_unicode_returns method. on MySQLDialect_mysqldb. I think is not used after https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1946, and is calling to a super that I think doesnt exist anymore.
- delete _set_isolation_level, is unused, and has a super that doesnt exists anymore: 
https://github.com/sqlalchemy/sqlalchemy/commit/af1b91626f63e00e11d07ad378d23198abc7f91f
- Separate _bitmap in enumerated to _bitmap and _inversed_bitmap to have type safety
- Add inheritance from log.Identified to MySQLTableDefinitionParser, because has the decorathor @log.class_logger

Related to issue :
https://github.com/sqlalchemy/sqlalchemy/issues/6810

CaselIT commented on lib/sqlalchemy/engine/interfaces.py:
```diff
@@ -122,7 +122,7 @@ def close(self) -> None: ...
     def commit(self) -> None: ...
 
-    def cursor(self) -> DBAPICursor: ...
+    def cursor(self, *args: Any, **kwargs: Any) -> DBAPICursor: ...
```
were the missing types causing some type issues?

Polandia94 replied:
yes, on `mysql.base.create_server_side_cursor`, cursor() is called with parmeter `self.dialect._sscursor`.

CaselIT commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -3512,7 +3519,7 @@ def __getitem__(self, index):
     def __getattr__(self, attr):
         item = getattr(self.rowproxy, attr)
         if isinstance(item, _array):
```
We can probably just remove this if (and the above one), considering that no one reported the error and 3.9 was more than 4 years ago.

CaselIT reviewed (commented):
feel free to ask if you have any question

CaselIT commented:
Sorry, wanted to make the ci run but somehow made the PR no longer draft. Feel free to set again as draft

CaselIT commented on lib/sqlalchemy/dialects/mysql/mysqldb.py:
```diff
@@ -164,43 +175,24 @@ def on_connect(conn):
         if context is not None:
             context._rowcount = rowcount
 
-    def _check_unicode_returns(self, connection):
```
uh. no one was calling this? I guess it was a leftover of when it was removed

CaselIT replied:
cc @zzzeek

CaselIT commented:
thanks for the effort! Just so you know if you want a maintainer to take over just say so :)

nightblure commented on lib/sqlalchemy/connectors/asyncio.py:
```diff
@@ -40,7 +40,7 @@ async def close(self) -> None: ...
     async def commit(self) -> None: ...
 
-    def cursor(self) -> AsyncIODBAPICursor: ...
+    def cursor(self, *args, **kwargs) -> AsyncIODBAPICursor: ...
```
What do you think about adding an annotation?
`*args: Any, **kwargs: Any`

Polandia94 commented:
> @Polandia94 this is starting to become a bit too big. Can you split it by any change? At the moment a logic split could be everything not in dialect/mysql followed by what's in dialect/mysql.
> 
> note that's fine also if you want to keep doing everything here then split it in multiple pr later.

I expect to end this work in arround 2 or 3 weeks, and after that I will divide this in a couple of PRs. Probablly 1 for not in dialect, and 2 or 3 for the dialect, so we can review better.

zzzeek commented:
looks pretty substantial.   i didnt see any red flags yet...

CaselIT commented:
> > @Polandia94 this is starting to become a bit too big. Can you split it by any change? At the moment a logic split could be everything not in dialect/mysql followed by what's in dialect/mysql.
> > note that's fine also if you want to keep doing everything here then split it in multiple pr later.
> 
> I expect to end this work in arround 2 or 3 weeks, and after that I will divide this in a couple of PRs. Probablly 1 for not in dialect, and 2 or 3 for the dialect, so we can review better.

Looks like a plan! Ping us when you want to proceed!

Polandia94 commented:
This is ready, so i will start to creating shorter PRs, with parts of this.
I added #  type: ignore on some imports and ignore the zimports pre-commit that delete that comments.
Also create a zimports PR about that, https://github.com/sqlalchemyorg/zimports/pull/43

Dreamsorcerer commented:
> you can ignore untyped module erros using a module level mypy comment

I'd note you can do this is the mypy.ini config, as we do (which gives us a central place to check for modules that still need to be typed):
https://github.com/aio-libs/aiohttp/blob/c738a83d5ad0c270df94b8f93a206f0c348896e5/.mypy.ini#L36-L37

> also note that we try to use specific type ignores using for example `# type: ignore[import-untyped]` instead of a general `# type: ignore`

This can also be enforced with `enable_error_code = ignore-without-code`.

CaselIT commented:
> This can also be enforced with enable_error_code = ignore-without-code.

seems useful, but a current run says

> Found 341 errors in 64 files (checked 248 source files)

so at the very least it will not be in this PR

Polandia94 commented:
@CaselIT I think this i ready for review.
I was waiting to https://github.com/sqlalchemy/sqlalchemy/pull/12231

dlax commented on lib/sqlalchemy/dialects/mysql/json.py:
```diff
@@ -34,43 +42,29 @@ class JSON(sqltypes.JSON):
-        return process
-
-    def literal_processor(self, dialect):
-        super_proc = self.string_literal_processor(dialect)
-
-        def process(value):
-            value = self._format_value(value)
-            if super_proc:
-                value = super_proc(value)
-            return value
+    def bind_processor(self, dialect: Dialect) -> "_BindProcessorType[Any]":
+        raise NotImplementedError()
 
-        return process
+    def literal_processor(
+        self, dialect: Dialect
+    ) -> "_LiteralProcessorType[Any]":
+        raise NotImplementedError()
```
These methods weren't actually used?

CaselIT replied:
it seems strange that they aren't used.

Polandia94 replied:
I really dont remember why I removed this methods, probablly a mstake

dlax commented on lib/sqlalchemy/dialects/mysql/mariadbconnector.py:
```diff
@@ -201,19 +223,21 @@ def create_connect_args(self, url):
 
-    def get_isolation_level_values(self, dbapi_connection):
+    def get_isolation_level_values(
+        self, dbapi_conn: "DBAPIConnection"
+    ) -> tuple["IsolationLevel", ...]:
```
```suggestion
        self, dbapi_conn: DBAPIConnection) -> Sequence[IsolationLevel];
```
(quotes removed, return type consistent with parent)

dlax commented on lib/sqlalchemy/dialects/mysql/mariadbconnector.py:
```diff
@@ -222,21 +246,23 @@ def get_isolation_level_values(self, dbapi_connection):
 
-    def set_isolation_level(self, connection, level):
+    def set_isolation_level(
+        self, dbapi_connection: "DBAPIConnection", level: "IsolationLevel"
```
```suggestion
        self, dbapi_connection: DBAPIConnection, level: IsolationLevel
```
?

dlax commented on lib/sqlalchemy/dialects/mysql/mysqldb.py:
```diff
@@ -285,7 +283,9 @@ def _detect_charset(self, connection):
             return cset_name()
 
-    def get_isolation_level_values(self, dbapi_connection):
+    def get_isolation_level_values(
+        self, dbapi_conn: "DBAPIConnection"
+    ) -> tuple["IsolationLevel", ...]:
```
```suggestion
    def get_isolation_level_values(
        self, dbapi_conn: DBAPIConnection
    ) -> Sequence[IsolationLevel]:
```

dlax commented on lib/sqlalchemy/dialects/mysql/mysqldb.py:
```diff
@@ -294,12 +294,14 @@ def get_isolation_level_values(self, dbapi_connection):
         )
 
-    def set_isolation_level(self, dbapi_connection, level):
+    def set_isolation_level(  # type: ignore[override]
+        self, dbapi_connection: "MySQLdb.Connection", level: "IsolationLevel"
+    ) -> None:
```
Alternatively:
```suggestion
    def set_isolation_level(
        self, dbapi_connection: DBAPIConnection, level: IsolationLevel
    ) -> None:
        if TYPE_CHECKING:
            assert isinstance(dbapi_connection, MySQLdb.Connection)
```
would remove the need for extra `# type: ignore` while keeping the signature consistent with parent.

Polandia94 replied:
this function call `dbapi_connection.autocommit(True)`, that is from` MySQLdb.Connection` but not from `DBAPIConnection`. We can go but this approach, but would be need to add type ignore in that calls.

dlax commented on lib/sqlalchemy/dialects/mysql/pyodbc.py:
```diff
@@ -55,23 +61,32 @@
-        def process(value):
+    def result_processor(
+        self, dialect: Dialect, coltype: object
+    ) -> "_ResultProcessorType[datetime.time]":
+        def process(value: Any) -> "datetime.time | None":
```
quotes may be removed (provided `Union` is used instead of `|`)

dlax commented on lib/sqlalchemy/dialects/mysql/reflection.py:
```diff
@@ -18,31 +25,40 @@
 
 @log.class_logger
-class MySQLTableDefinitionParser:
+class MySQLTableDefinitionParser(log.Identified):
```
It may be useful to explain in commit message (PR description) why the inheritance got changed.

CaselIT replied:
This is a typing PR, any change that's not insignificant should be taken out.

dlax reviewed (commented):
Left some suggestions inline; I have skipped reviewing `lib/sqlalchemy/dialects/mysql/types.py` changes (lack of time for now).

Polandia94 commented on lib/sqlalchemy/dialects/mysql/aiomysql.py:
```diff
@@ -137,39 +162,46 @@ class MySQLDialect_aiomysql(MySQLDialect_pymysql):
+    # _translate_args should not be defined here, is only for super class.
+    def create_connect_args(
+        self, url: URL, _translate_args: Any = None
+    ) -> "ConnectArgsType":
```
it should also receive kwargs. Not sure what is better

CaselIT commented:
> @CaselIT I think this i ready for review. I was waiting to #12231

sorry but this is already an huge change, don't change also the implemented code. Things like adding additional base classes to the dialect or similar are out of scope for this PR.

These parts should be removed and if you are interested another PR can be opened for them.

Thanks

Polandia94 commented on lib/sqlalchemy/dialects/mysql/asyncmy.py:
```diff
@@ -57,52 +79,60 @@ class AsyncAdapt_asyncmy_connection(AsyncAdapt_dbapi_connection):
-    def character_set_name(self):
-        return self._connection.character_set_name()
+    def character_set_name(self) -> Optional[str]:
+        return self._connection.character_set_name()  # type: ignore[no-any-return]  # noqa: E501
```
connection is of type AsyncmyConnection(that is really the Connection class of Asyncmy library), not AsyncIODBAPIConnection.

Polandia94 commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -2157,28 +2310,36 @@ def visit_drop_constraint(self, drop, **kw):
+    def visit_set_table_comment(
+        self, create: "ddl.SetTableComment", **kw: Any
+    ) -> str:
         return "ALTER TABLE %s COMMENT %s" % (
             self.preparer.format_table(create.element),
             self.sql_compiler.render_literal_value(
                 create.element.comment, sqltypes.String()
             ),
         )
 
-    def visit_drop_table_comment(self, create, **kw):
+    def visit_drop_table_comment(
+        self, drop: "ddl.DropTableComment", **kw: Any
+    ) -> str:
         return "ALTER TABLE %s COMMENT ''" % (
-            self.preparer.format_table(create.element)
+            self.preparer.format_table(drop.element)
         )
 
-    def visit_set_column_comment(self, create, **kw):
+    def visit_set_column_comment(
+        self, create: "ddl.SetColumnComment", **kw: Any
+    ) -> str:
```
This is just to have the same names that superclass sql.compiler.DDLCompiler

Polandia94 commented:
> > @CaselIT I think this i ready for review. I was waiting to #12231
> 
> sorry but this is already an huge change, don't change also the implemented code. Things like adding additional base classes to the dialect or similar are out of scope for this PR.
> 
> These parts should be removed and if you are interested another PR can be opened for them.
> 
> Thanks

Sorry, I think i removed all changes to implemented code.

CaselIT commented:
> Sorry, I think i removed all changes to implemented code.

ok thanks!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 29879417fad1253d4cf1e54425fcb63e5367fdb8 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 29879417fad1253d4cf1e54425fcb63e5367fdb8: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/connectors/aioodbc.py:
```diff
@@ -138,7 +140,9 @@ class aiodbcConnector(PyODBCConnector):
     supports_statement_cache = True
 
-    supports_server_side_cursors = True
+    supports_server_side_cursors: Union[generic_fn_descriptor[bool], bool] = (
```
**Federico Caselli** (CaselIT) wrote:

strange that you have to re-state the typing. I thought it was not needed since it's already defined in the base class

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/__init__.py:
```diff
@@ -39,7 +40,7 @@ def _auto_fn(name: str) -> Optional[Callable[[], Type[Dialect]]]:
             # they would just publish all the entrypoints, which would actually
             # look much nicer.
-            module = __import__(
+            module: "mariadb" = __import__(  # type: ignore[valid-type]
```
**Federico Caselli** (CaselIT) wrote:

let's just use any since this does not work anyway

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/enumerated.py:
```diff
@@ -4,26 +4,39 @@
     native_enum = True
 
-    def __init__(self, *enums, **kw):
+    def __init__(self, *enums: Union[str, Enum], **kw: Any):
```
**Federico Caselli** (CaselIT) wrote:

note that all init you type should by typed with returning `-> None:`
usually mypy complained with "partially typed function" not sure why it's not doing that here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Dreamsorcerer replied:
An oddity from when typing was originally introduced, is that the return type on `__init__` is considered optional and will be inferred as None. There must be a type annotation _somewhere_ though for it to be treated as typed, so if there are no parameters, a return annotation needs to be added. That historical oddity is why mypy doesn't error here.

sqla-tester commented on lib/sqlalchemy/dialects/mysql/expression.py:
```diff
@@ -4,9 +4,10 @@
 # This module is part of SQLAlchemy and is released under
 # the MIT License: https://www.opensource.org/licenses/mit-license.php
-# mypy: ignore-errors
 
```
**Federico Caselli** (CaselIT) wrote:

strange that the future-annotation flake8 warning does not came up here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Polandia94 replied:
It's because there is no use of any sintax that require, but probablly better to add it

sqla-tester replied:
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/json.py:
```diff
@@ -34,43 +42,45 @@ class JSON(sqltypes.JSON):
         else:
             value = '$."%s"' % value
-        return value
+        return value  # type: ignore[no-any-return]
```
**Federico Caselli** (CaselIT) wrote:

lets use a variable named something else instead

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/mariadb.py:
```diff
@@ -46,23 +58,23 @@ def __init__(self, as_uuid: bool = True, native_uuid: bool = True):
         return True
 
-    def bind_processor(self, dialect):
+    def bind_processor(self, dialect: "MariaDBDialect") -> Optional[_BindProcessorType[_UUID_RETURN]]:  # type: ignore[override] # noqa: E501
```
**Federico Caselli** (CaselIT) wrote:

can remove the `""` in the type def here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/sql/ddl.py:
```diff
@@ -32,6 +32,7 @@
 from .base import Executable
 from .base import SchemaVisitor
 from .elements import ClauseElement
+from .schema import Column
```
**Federico Caselli** (CaselIT) wrote:

please revert the change to column, let's keep in type checking

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/sql/ddl.py:
```diff
@@ -715,6 +715,7 @@ class CreateIndex(_CreateBase["Index"]):
     """Represent a CREATE INDEX statement."""
 
     __visit_name__ = "create_index"
+    element: Index
```
**Federico Caselli** (CaselIT) wrote:

how about we instead add `element: _SI` to _CreateDropBase?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/sql/sqltypes.py:
```diff
@@ -1494,7 +1494,9 @@ def _enum_init(self, enums: _EnumTupleArg, kw: Dict[str, Any]) -> None:
         self.native_enum = kw.pop("native_enum", True)
         self.create_constraint = kw.pop("create_constraint", False)
-        self.values_callable = kw.pop("values_callable", None)
+        self.values_callable: Optional[
```
**Federico Caselli** (CaselIT) wrote:

please define this on the class. same also below

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/sql/sqltypes.py:
```diff
@@ -3512,6 +3518,7 @@ class BINARY(_Binary):
 class VARBINARY(_Binary):
     """The SQL VARBINARY type."""
 
+    length: Optional[int]
```
**Federico Caselli** (CaselIT) wrote:

_Binary already defines length so this can be removed

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on pyproject.toml:
```diff
@@ -46,7 +46,8 @@ Discussions = "https://github.com/sqlalchemy/sqlalchemy/discussions"
     "mypy >= 1.7",
-    "types-greenlet >= 2"
+    "types-greenlet >= 2",
+    "types-PyMySQL >=1.1"
```
**Federico Caselli** (CaselIT) wrote:

this does not seem complete judging by tox

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Polandia94 replied:
added 
types-mysqlclient
the no tipyng are in the other dependencies(also added cymysql that was nowhere).

sqla-tester commented on tox.ini:
```diff
@@ -191,6 +191,14 @@ deps=
+     mysql-connector-python
+     cymysql
+     asyncmy
+     aiomysql
```
**Federico Caselli** (CaselIT) wrote:

it's unfortunate that we have install so many additional types here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

CaselIT replied:
cc @zzzeek

sqla-tester replied:
**Federico Caselli** (CaselIT) wrote:

We preferred solving this by using ignores in mypy, since otherwise for each dialect we would need to install a stub and that becomes a mess very fast

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

sorry for the delay. looks mostly ok

not yet finished locking all files

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/aiomysql.py:
```diff
@@ -137,39 +162,46 @@ class MySQLDialect_aiomysql(MySQLDialect_pymysql):
-    def create_connect_args(self, url):
+    # _translate_args should not be defined here, is only for super class.
+    def create_connect_args(
+        self, url: URL, _translate_args: Any = None
```
**Federico Caselli** (CaselIT) wrote:

let's use type ignore intead

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/asyncmy.py:
```diff
@@ -138,25 +168,30 @@ class MySQLDialect_asyncmy(MySQLDialect_pymysql):
-    def create_connect_args(self, url):
+    # _translate_args should not be defined here, is only for super class.
+    def create_connect_args(
+        self, url: URL, _translate_args: Any = None
```
**Federico Caselli** (CaselIT) wrote:

same as the other note

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1346,17 +1406,17 @@ def _render_json_extract_from_binary(self, binary, operator, **kw):
             )
         elif binary.type._type_affinity in (sqltypes.Numeric, sqltypes.Float):
             if (
```
**Federico Caselli** (CaselIT) wrote:

maybe we could more simply do `bin_type_: sqltypes.Numeric = binary.type # type: ignore` and then use `bin_type_`?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Polandia94 replied:
I will change to
binary.type = cast(sqltypes.Numeric, binary.type)
inside that branch

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1434,24 +1500,28 @@ def visit_on_duplicate_key_update(self, on_duplicate, **kw):
-                    return obj._with_binary_element_type(column.type)
+                    element = element._clone()
+                    element.type = column.type
+                    return element
```
**Federico Caselli** (CaselIT) wrote:

please restore the previous code

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -2440,40 +2602,43 @@ def visit_NCHAR(self, type_, **kw):
             return self._extend_string(type_, {"national": True}, "CHAR")
 
-    def visit_UUID(self, type_, **kw):
+    def visit_UUID(self, type_: UUID[sqltypes._UUID_RETURN], **kw: Any) -> str:  # type: ignore[override]  # NOQA: E501
```
**Federico Caselli** (CaselIT) wrote:

it's likely fine to use `UUID[Any]`, buy this is also fine

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -3547,7 +3900,7 @@ class _DecodingRow:
     # seem to come up in DDL queries.
 
-    _encoding_compat = {
+    _encoding_compat: dict[Optional[str], str] = {
```
**Federico Caselli** (CaselIT) wrote:

why optional for the key?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Polandia94 replied:
this is used below on 
`self._encoding_compat.get(charset, charset)
and charset could be None. 

would add a comment, but we can also use dict[str,str] and add a type ignore below

https://github.com/python/typeshed/issues/6597

sqla-tester commented on lib/sqlalchemy/dialects/mysql/mysqlconnector.py:
```diff
@@ -120,21 +151,28 @@ class MySQLDialect_mysqlconnector(MySQLDialect):
     execution_ctx_cls = MySQLExecutionContext_mysqlconnector
 
-    preparer = MySQLIdentifierPreparer_mysqlconnector
+    execution_ctx_cls = MySQLExecutionContext_mysqlconnector
```
**Federico Caselli** (CaselIT) wrote:

this line got duplicated

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/reflection.py:
```diff
@@ -18,31 +25,40 @@
+    ) -> ReflectedState:
         state = ReflectedState()
-        state.charset = charset
+        state.charset = charset  # type: ignore[attr-defined]
```
**Federico Caselli** (CaselIT) wrote:

let's define it on the class 
```py
class ReflectedState:
  ...
  charset: Optional[str]
```

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/reflection.py:
```diff
@@ -379,7 +407,9 @@ def _describe_to_create(self, table_name, columns):
-    def _parse_keyexprs(self, identifiers):
+    def _parse_keyexprs(
+        self, identifiers: str
+    ) -> list[tuple[Any, Optional[int], Any]]:
```
**Federico Caselli** (CaselIT) wrote:

isn't this `list[tuple[str, Optional[int], str]]`?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

should looked at everything now. I've noted just a few minor things

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Polandia94 commented:
@CaselIT I think i changed all comments hopefully. Not sure why are not running the checks, i guess this commits has to be ported to gerrit somehow?
Tomorrow will check again the comments and mark as solved all the solved

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision b877ccab90864cd8d514809e8ad1a6cab7b20494 of this pull request into gerrit so we can run tests and reviews and stuff

Polandia94 committed:
Type mysql dialect

Tipyng of mysql dialect.

possible breaking changes:

- on `visit_sequence`, `visit_true` and `visit_false` I changed the param names to be the same tht in the super class.
- Removed  `tostring` on `array`. `tostring` is deprecated since Python 3.2 and removed in Python3.9 so i guess we can remove that option altogheter.
- create_connect_args on aiomysql, added an unused param _translate_args, only to have same signature as super class.
-  delete _check_unicode_returns method. on MySQLDialect_mysqldb. I think is not used after https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1946, and is calling to a super that I think doesnt exist anymore.
- delete _set_isolation_level, is unused, and has a super that doesnt exists anymore:
https://github.com/sqlalchemy/sqlalchemy/commit/af1b91626f63e00e11d07ad378d23198abc7f91f
- Separate _bitmap in enumerated to _bitmap and _inversed_bitmap to have type safety
- Add inheritance from log.Identified to MySQLTableDefinitionParser, because has the decorathor @log.class_logger

Related to issue :
https://github.com/sqlalchemy/sqlalchemy/issues/6810

Closes: #12164
Pull-request: https://github.com/sqlalchemy/sqlalchemy/pull/12164
Pull-request-sha: b877ccab90864cd8d514809e8ad1a6cab7b20494

Change-Id: I37bd98049ff1a64d58e9490b0e5e2ea764dd1f73

sqla-tester commented:
Patchset b877ccab90864cd8d514809e8ad1a6cab7b20494 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/__init__.py:
```diff
@@ -39,7 +40,7 @@ def _auto_fn(name: str) -> Optional[Callable[[], Type[Dialect]]]:
             # they would just publish all the entrypoints, which would actually
             # look much nicer.
-            module = __import__(
+            module: Any = __import__(
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

thanks looks mostly fine. I'll do a final pass

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

* lib/sqlalchemy/connectors/aioodbc.py (line 143): Done
* lib/sqlalchemy/sql/ddl.py (line 35): Done
* lib/sqlalchemy/sql/ddl.py (line 718): Done
* lib/sqlalchemy/sql/sqltypes.py (line 1497): Done
* lib/sqlalchemy/sql/sqltypes.py (line 3521): Done

sqla-tester commented on lib/sqlalchemy/dialects/mysql/base.py:
```diff
@@ -1345,6 +1405,7 @@ def _render_json_extract_from_binary(self, binary, operator, **kw):
                 )
             )
         elif binary.type._type_affinity in (sqltypes.Numeric, sqltypes.Float):
+            binary.type = cast(sqltypes.Numeric[Any], binary.type)
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on pyproject.toml:
```diff
@@ -46,7 +46,9 @@ Discussions = "https://github.com/sqlalchemy/sqlalchemy/discussions"
     "mypy >= 1.7",
-    "types-greenlet >= 2"
+    "types-greenlet >= 2",
+    "types-PyMySQL >=1.1",
```
**Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

let's see if it still passes typecheck

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Polandia94 commented:
@CaselIT I think with your last patchset all comments are solved and is passing the tests. Let me know if you need that i change something on the PR

CaselIT commented:
yes, it should be done. I'm waiting for a review from mike. We should be able to merge it soonish

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

I planned to do that after it was done. Will do later today

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester commented on lib/sqlalchemy/dialects/mysql/mysqldb.py:
```diff
@@ -145,17 +161,17 @@ def supports_server_side_cursors(self):
         return __import__("MySQLdb")
 
-    def on_connect(self):
+    def on_connect(self) -> Callable[["MySQLdb.Connection"], None]:
```
**Michael Bayer** (zzzeek) wrote:

i dont think this is the approach we should be taking here, trying to get a handle on specific types inside the DBAPIs that typically don't have stubs.  We have protocols for DBAPI connection, cursor DBAPIConnection, DBAPICursor and thats what should be used here.  I also note that if we go back to engine/interfaces engine/default the callable here has Any in it, that should be DBAPIConnection etc.

here's how this can be done for this method, for example, including interfaces / default.  it passes mypy 100%


```
diff --git a/lib/sqlalchemy/dialects/mysql/mysqldb.py b/lib/sqlalchemy/dialects/mysql/mysqldb.py
index ae955389e..a6780f94b 100644
--- a/lib/sqlalchemy/dialects/mysql/mysqldb.py
+++ b/lib/sqlalchemy/dialects/mysql/mysqldb.py
@@ -89,7 +89,7 @@ from __future__ import annotations
 
 import re
 from types import ModuleType
-from typing import Any
+from typing import Any, cast
 from typing import Callable
 from typing import Iterable
 from typing import Literal
@@ -164,14 +164,14 @@ class MySQLDialect_mysqldb(MySQLDialect):
     def import_dbapi(cls) -> ModuleType:
         return __import__("MySQLdb")
 
-    def on_connect(self) -> Callable[[MySQLdb.Connection], None]:
+    def on_connect(self) -> Callable[[DBAPIConnection], None]:
         super_ = super().on_connect()
 
-        def on_connect(conn: MySQLdb.Connection) -> None:
+        def on_connect(conn: DBAPIConnection) -> None:
             if super_ is not None:
                 super_(conn)
 
-            charset_name = conn.character_set_name()
+            charset_name = cast(MySQLdb.Connection, conn).character_set_name()
 
             if charset_name is not None:
                 cursor = conn.cursor()
diff --git a/lib/sqlalchemy/engine/default.py b/lib/sqlalchemy/engine/default.py
index 45c00fe0e..ac3048b6b 100644
--- a/lib/sqlalchemy/engine/default.py
+++ b/lib/sqlalchemy/engine/default.py
@@ -563,7 +563,7 @@ class DefaultDialect(Dialect):
                 % (self.label_length, self.max_identifier_length)
             )
 
-    def on_connect(self) -> Optional[Callable[[Any], None]]:
+    def on_connect(self) -> Optional[Callable[[DBAPIConnection], None]]:
         # inherits the docstring from interfaces.Dialect.on_connect
         return None
 
diff --git a/lib/sqlalchemy/engine/interfaces.py b/lib/sqlalchemy/engine/interfaces.py
index 556e49d7f..7a6c4005e 100644
--- a/lib/sqlalchemy/engine/interfaces.py
+++ b/lib/sqlalchemy/engine/interfaces.py
@@ -2307,7 +2307,7 @@ class Dialect(EventTarget):
         """
         return self.on_connect()
 
-    def on_connect(self) -> Optional[Callable[[Any], None]]:
+    def on_connect(self) -> Optional[Callable[[DBAPIConnection], None]]:
         """return a callable which sets up a newly created DBAPI connection.
 
         The callable should accept a single argument "conn" which is the

```

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

if this change is local to gerrit now and we can iterate here, I want to go through and change most of the use of third party names like MySQLdb.Connection and replace with protocols DBAPIConnection, etc., that's what these are for.  we can use cast() as needed when we need a specific method, but otherwise the patch here is very dependent on lots and lots of third party classes that typically aren't typed, could change names (and have awkward names anyway), the code is not as easy to follow as the main advantage to typing is understanding the signature of functions.  using consistent types all the way through using protocols will make this much more clear and the code is more robust for subclassing, etc.

i can do this locally if we are in a gerrit iteration now.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

zzzeek commented:
I see you did those changes thanks! ( I was going to do them) - is this PR in sync with the gerrit we have?  I can re-push from here into gerrit but I dont want to lose changes that @CaselIT may have made in gerrit only

sqla-tester commented:
**Pablo Nicolás Estevez** (Polandia94) wrote:

I created a new commit on github, with this approach to pyodbc, pymysql and aiomysql, is in the PR branch.

I find it simplier to check the diff on github here:
https://github.com/Polandia94/sqlalchemy/pull/1/files

If this is the approach you are looking for I can change the next of the modules next week.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

Polandia94 committed:
merge branch

Polandia94 committed:
fix test on PR

sqla-tester commented:
**Pablo Nicolás Estevez** (Polandia94) wrote:

hi, I don't know how to use Gerrit. 
I created a new cmmit with this changes over the last patch and pushed to my PR on github.
https://github.com/sqlalchemy/sqlalchemy/pull/12164

I have to froce-push to reconcile history, left the all history on https://github.com/Polandia94/sqlalchemy/tree/type-mysql-backup

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 545e2c39d5ee4f3938111b26e098fa2aa2b6e800 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 545e2c39d5ee4f3938111b26e098fa2aa2b6e800 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5875

* lib/sqlalchemy/dialects/mysql/base.py (line 3444): it's likely fine since it's a string anyway due to future-annotation but it's likely better to restore the uppercase if we do another change

sqla-tester reviewed (commented):
**Michael Bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5875

* lib/sqlalchemy/dialects/mysql/base.py (line 3444): OK, there's tons of `lowercase[foo]` in this particular cherry pick, none of it causes a problem because we dont run mypy under 3.9.   will try to get most of them

OK I changed a few hundred of them

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5829 has been **merged**. Congratulations! :)

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5875 has been **merged**. Congratulations! :)
